qutebrowser/browser/webkit/downloads.py

```diff
@@ -926,19 +936,51 @@ def fetch(self, reply, *, fileobj=None, filename=None, auto_remove=False,
 
         # User doesn't want to be asked, so just use the download_dir
         if filename is not None:
-            download.set_filename(filename)
+            self.set_filename_for_download(download, suggested_filename,
+                                           filename)
             return download
 
         # Ask the user for a filename
         self._postprocess_question(q)
-        q.answered.connect(download.set_filename)
+        q.answered.connect(
+            functools.partial(self.set_filename_for_download, download,
+                              suggested_filename))
         q.cancelled.connect(download.cancel)
         download.cancelled.connect(q.abort)
         download.error.connect(q.abort)
         q.ask()
 
         return download
 
+    def set_filename_for_download(self, download, suggested_filename,
+                                  filename):
+        """Set the filename for a given download.
+
+        This correctly handles the case where filename = OPEN_DOWNLOAD.
+
+        Args:
+            download: The download to set the filename for.
+            suggested_filename: The suggested filename.
+            filename: The filename as string or usertypes.OPEN_DOWNLOAD.
+        """
+        if filename is not usertypes.OPEN_DOWNLOAD:
+            download.set_filename(filename)
+            return
+        # Find the next free filename without causing a race condition
```

Comment:
Couldn't we just use [`tempfile.NamedTemporaryFile`](https://docs.python.org/3.5/library/tempfile.html#tempfile.NamedTemporaryFile) with `prefix` and `suffix` set appropriately?